tests/performance/scripts/entrypoint.sh

```diff
@@ -88,19 +89,9 @@ chmod 777 workspace output
 
 cd workspace
 
-# Download the package for the version we are going to test.
-# A temporary solution for migrating into PRs directory
-for prefix in "$S3_URL/PRs" "$S3_URL";
-do
-    if curl_with_retry "$prefix/$PR_TO_TEST/$SHA_TO_TEST/$BUILD_NAME/performance.tar.zst"
-    then
-        right_path="$prefix/$PR_TO_TEST/$SHA_TO_TEST/$BUILD_NAME/performance.tar.zst"
-        break
-    fi
-done
-
-mkdir right
-wget -nv -nd -c "$right_path" -O- | tar -C right --no-same-owner --strip-components=1 --zstd --extract --verbose
+[ ! -e "/artifacts/performance.tar.zst" ] && echo "ERROR: performance.tar.zst not found" && exit 1
+mkdir -p right
+tar -xf "/artifacts/performance.tar.zst" -C right --no-same-owner --strip-components=1 --zstd --extract --verbose
```

Comment:
Am I missing something or because of this change you cannot test new performance tests? (since it assume that patched-right is the same as upstream-left) And not to mention changes in `compare.sh` and other scripts

Reply:
Thank you! You seem to point out an interesting use case with performance tests that I missed during this ci enhancement. I think I understand the problem you're talking about, but anyway it would be helpful if you could share a PR or something else where the problem manifests itself so I can make sure I know all the details.

I'm guessing that you encounter the situation that tests are run with some outdated performance artifact without new tests you have added.
Due to the changes I made in CI, build jobs could be skipped if nothing has changed in the source files. And then the performance test will re-use some old performance artifact (which was made on the same sources) and this could lead to this problem because no new tests there.

If my understanding is correct, there is a quick way to solve this problem in your PR: change something in the source file so that the build job is not skipped.

If you have links to share, that would be helpful. I will check and figure out how to fix this.

Reply:
> I think I understand the problem you're talking about, but anyway it would be helpful if you could share a PR or something else where the problem manifests itself so I can make sure I know all the details.

- you can try to add  new perf test and likely it will not be tested on CI
- modifications to `compare.sh` - https://github.com/ClickHouse/ClickHouse/pull/59287

>I'm guessing that you encounter the situation that tests are run with some outdated performance artifact without new tests you have added.

Not outdated, it simply uses upstream for both, while it should use upstream **and** patched version (from PR), otherwise it will not have any changes!

Reply:
There are even more oddities.

Actually this artifacts will be downloaded in python - https://github.com/ClickHouse/ClickHouse/actions/runs/7678971289/job/20929742660 (for commit https://github.com/ClickHouse/ClickHouse/commit/661d4a827b38f1a60840fb9178a76dc13bcccdf3):

```
2024-01-27T14:13:22.5133828Z INFO:root:Will download performance.tar.zst to /home/ubuntu/actions-runner/_work/_temp/test
2024-01-27T14:13:22.5134977Z INFO:root:Downloading from https://s3.amazonaws.com/clickhouse-builds/PRs/0/661d4a827b38f1a60840fb9178a76dc13bcccdf3/package_release/performance.tar.zst to temp path /home/ubuntu/actions-runner/_work/_temp/test/performance.tar.zst
2024-01-27T14:13:22.5136366Z INFO:root:Getting URL with 1 tries and sleep 3 in between: https://s3.amazonaws.com/clickhouse-builds/PRs/0/661d4a827b38f1a60840fb9178a76dc13bcccdf3/package_release/performance.tar.zst
2024-01-27T14:13:22.6730355Z INFO:root:Content length is 867927141 bytes
2024-01-27T14:13:44.0817743Z INFO:root:Downloading finished
2024-01-27T14:13:44.0857017Z INFO:root:Going to run command docker run --privileged --volume=/home/ubuntu/actions-runner/_work/_temp/test/result:/workspace --volume=/home/ubuntu/actions-runner/_work/_temp/test/result:/output --volume=/home/ubuntu/actions-runner/_work/ClickHouse/ClickHouse/tests:/usr/share/clickhouse-test --volume=/home/ubuntu/actions-runner/_work/_temp/test:/artifacts --cap-add syslog --cap-add sys_admin --cap-add sys_rawio -e CHECK_START_TIME='2024-01-27 14:12:48' -e CHECK_NAME='Performance Comparison' -e INSTANCE_TYPE='m7i-flex.4xlarge' -e INSTANCE_ID='i-0ea15b9a0ac2178be' -e PR_TO_TEST=0 -e SHA_TO_TEST=661d4a827b38f1a60840fb9178a76dc13bcccdf3  -e S3_URL=https://s3.amazonaws.com/clickhouse-builds -e BUILD_NAME=package_release -e CHPC_ADD_REPORT_LINKS="<a href=https://github.com/ClickHouse/ClickHouse/actions/runs/7678971289>Job (actions)</a> <a href=https://github.com/ClickHouse/ClickHouse/commit/661d4a827b38f1a60840fb9178a76dc13bcccdf3>Tested commit</a>" -e CHPC_TEST_RUN_BY_HASH_TOTAL=4 -e CHPC_TEST_RUN_BY_HASH_NUM=0 -e CLICKHOUSE_PERFORMANCE_COMPARISON_DATABASE_URL -e CLICKHOUSE_PERFORMANCE_COMPARISON_DATABASE_USER -e CLICKHOUSE_PERFORMANCE_COMPARISON_DATABASE_USER_PASSWORD -e CLICKHOUSE_PERFORMANCE_COMPARISON_CHECK_NAME -e CLICKHOUSE_PERFORMANCE_COMPARISON_CHECK_NAME_PREFIX clickhouse/performance-comparison:adc182c3de19
...
2024-01-27T14:14:13.4816209Z + '[' '!' -e /artifacts/performance.tar.zst ']'
2024-01-27T14:14:13.4816524Z + mkdir -p right
2024-01-27T14:14:13.4826395Z + tar -xf /artifacts/performance.tar.zst -C right --no-same-owner --strip-components=1 --zstd --extract --verbose
```

But for instance for this commit
https://github.com/ClickHouse/ClickHouse/commit/c7289f69edca4f3ae7cdcbb620b4ffc0cfde1fc7 (this is the merged PR, that had been mentioned above)

It downloads wrong archive:

```
2024-01-27T22:08:31.2612725Z INFO:root:Will download performance.tar.zst to /home/ubuntu/actions-runner/_work/_temp/test
2024-01-27T22:08:31.2614922Z INFO:root:Downloading from https://s3.amazonaws.com/clickhouse-builds/PRs/0/661d4a827b38f1a60840fb9178a76dc13bcccdf3/package_release/performance.tar.zst to temp path /home/ubuntu/actions-runner/_work/_temp/test/performance.tar.zst
```

And AFAICS for PRs it has upstream/master there, i.e. for https://github.com/ClickHouse/ClickHouse/pull/59287:

```
2024-01-27T12:32:24.0726233Z INFO:root:Found build report json build_report_master_package_release.json
2024-01-27T12:32:24.0727835Z INFO:root:Will download performance.tar.zst to /home/ubuntu/actions-runner/_work/_temp/test
2024-01-27T12:32:24.0759256Z ['https://s3.amazonaws.com/clickhouse-builds/24.1/875efc54b86c46f7ecb1d373627b6806da1873ef/package_release/clickhouse', 'https://s3.amazonaws.com/clickhouse-builds/24.1/875efc54b86c46f7ecb1d373627b6806da1873ef/package_release/clickhouse-client-24.1.1.1806-amd64.tgz', 'https://s3.amazonaws.com/clickhouse-builds/24.1/875efc54b86c46f7ecb1d373627b6806da1873ef/package_release/clickhouse-client-24.1.1.1806-amd64.tgz.sha512', 'https://s3.amazonaws.com/clickhouse-builds/24.1/875efc54b86c46f7ecb1d373627b6806da1873ef/package_release/clickhouse-client-24.1.1.1806.x86_64.rpm', 'https://s3.amazonaws.com/clickhouse-builds/24.1/875efc54b86c46f7ecb1d373627b6806da1873ef/package_release/clickhouse-client_24.1.1.1806_amd64.deb', 'https://s3.amazonaws.com/clickhouse-builds/24.1/875efc54b86c46f7ecb1d373627b6806da1873ef/package_release/clickhouse-common-static-24.1.1.1806-amd64.tgz', 'https://s3.amazonaws.com/clickhouse-builds/24.1/875efc54b86c46f7ecb1d373627b6806da1873ef/package_release/clickhouse-common-static-24.1.1.1806-amd64.tgz.sha512', 'https://s3.amazonaws.com/clickhouse-builds/24.1/875efc54b86c46f7ecb1d373627b6806da1873ef/package_release/clickhouse-common-static-24.1.1.1806.x86_64.rpm', 'https://s3.amazonaws.com/clickhouse-builds/24.1/875efc54b86c46f7ecb1d373627b6806da1873ef/package_release/clickhouse-common-static-dbg-24.1.1.1806-amd64.tgz', 'https://s3.amazonaws.com/clickhouse-builds/24.1/875efc54b86c46f7ecb1d373627b6806da1873ef/package_release/clickhouse-common-static-dbg-24.1.1.1806-amd64.tgz.sha512', 'https://s3.amazonaws.com/clickhouse-builds/24.1/875efc54b86c46f7ecb1d373627b6806da1873ef/package_release/clickhouse-common-static-dbg-24.1.1.1806.x86_64.rpm', 'https://s3.amazonaws.com/clickhouse-builds/24.1/875efc54b86c46f7ecb1d373627b6806da1873ef/package_release/clickhouse-common-static-dbg_24.1.1.1806_amd64.deb', 'https://s3.amazonaws.com/clickhouse-builds/24.1/875efc54b86c46f7ecb1d373627b6806da1873ef/package_release/clickhouse-common-static_24.1.1.1806_amd64.deb', 'https://s3.amazonaws.com/clickhouse-builds/24.1/875efc54b86c46f7ecb1d373627b6806da1873ef/package_release/clickhouse-diagnostics', 'https://s3.amazonaws.com/clickhouse-builds/24.1/875efc54b86c46f7ecb1d373627b6806da1873ef/package_release/clickhouse-keeper', 'https://s3.amazonaws.com/clickhouse-builds/24.1/875efc54b86c46f7ecb1d373627b6806da1873ef/package_release/clickhouse-keeper-24.1.1.1806-amd64.tgz', 'https://s3.amazonaws.com/clickhouse-builds/24.1/875efc54b86c46f7ecb1d373627b6806da1873ef/package_release/clickhouse-keeper-24.1.1.1806-amd64.tgz.sha512', 'https://s3.amazonaws.com/clickhouse-builds/24.1/875efc54b86c46f7ecb1d373627b6806da1873ef/package_release/clickhouse-keeper-24.1.1.1806.x86_64.rpm', 'https://s3.amazonaws.com/clickhouse-builds/24.1/875efc54b86c46f7ecb1d373627b6806da1873ef/package_release/clickhouse-keeper-dbg-24.1.1.1806-amd64.tgz', 'https://s3.amazonaws.com/clickhouse-builds/24.1/875efc54b86c46f7ecb1d373627b6806da1873ef/package_release/clickhouse-keeper-dbg-24.1.1.1806-amd64.tgz.sha512', 'https://s3.amazonaws.com/clickhouse-builds/24.1/875efc54b86c46f7ecb1d373627b6806da1873ef/package_release/clickhouse-keeper-dbg-24.1.1.1806.x86_64.rpm', 'https://s3.amazonaws.com/clickhouse-builds/24.1/875efc54b86c46f7ecb1d373627b6806da1873ef/package_release/clickhouse-keeper-dbg_24.1.1.1806_amd64.deb', 'https://s3.amazonaws.com/clickhouse-builds/24.1/875efc54b86c46f7ecb1d373627b6806da1873ef/package_release/clickhouse-keeper.time-trace', 'https://s3.amazonaws.com/clickhouse-builds/24.1/875efc54b86c46f7ecb1d373627b6806da1873ef/package_release/clickhouse-keeper_24.1.1.1806_amd64.deb', 'https://s3.amazonaws.com/clickhouse-builds/24.1/875efc54b86c46f7ecb1d373627b6806da1873ef/package_release/clickhouse-library-bridge', 'https://s3.amazonaws.com/clickhouse-builds/24.1/875efc54b86c46f7ecb1d373627b6806da1873ef/package_release/clickhouse-library-bridge.time-trace', 'https://s3.amazonaws.com/clickhouse-builds/24.1/875efc54b86c46f7ecb1d373627b6806da1873ef/package_release/clickhouse-odbc-bridge', 'https://s3.amazonaws.com/clickhouse-builds/24.1/875efc54b86c46f7ecb1d373627b6806da1873ef/package_release/clickhouse-odbc-bridge.time-trace', 'https://s3.amazonaws.com/clickhouse-builds/24.1/875efc54b86c46f7ecb1d373627b6806da1873ef/package_release/clickhouse-server-24.1.1.1806-amd64.tgz', 'https://s3.amazonaws.com/clickhouse-builds/24.1/875efc54b86c46f7ecb1d373627b6806da1873ef/package_release/clickhouse-server-24.1.1.1806-amd64.tgz.sha512', 'https://s3.amazonaws.com/clickhouse-builds/24.1/875efc54b86c46f7ecb1d373627b6806da1873ef/package_release/clickhouse-server-24.1.1.1806.x86_64.rpm', 'https://s3.amazonaws.com/clickhouse-builds/24.1/875efc54b86c46f7ecb1d373627b6806da1873ef/package_release/clickhouse-server_24.1.1.1806_amd64.deb', 'https://s3.amazonaws.com/clickhouse-builds/24.1/875efc54b86c46f7ecb1d373627b6806da1873ef/package_release/clickhouse-stripped', 'https://s3.amazonaws.com/clickhouse-builds/24.1/875efc54b86c46f7ecb1d373627b6806da1873ef/package_release/clickhouse.time-trace', 'https://s3.amazonaws.com/clickhouse-builds/PRs/0/875efc54b86c46f7ecb1d373627b6806da1873ef/package_release/performance.tar.zst']
2024-01-27T12:32:24.0781404Z INFO:root:Downloading from https://s3.amazonaws.com/clickhouse-builds/PRs/0/875efc54b86c46f7ecb1d373627b6806da1873ef/package_release/performance.tar.zst to temp path /home/ubuntu/actions-runner/_work/_temp/test/performance.tar.zst
2024-01-27T12:32:24.0783417Z INFO:root:Getting URL with 1 tries and sleep 3 in between: https://s3.amazonaws.com/clickhouse-builds/PRs/0/875efc54b86c46f7ecb1d373627b6806da1873ef/package_release/performance.tar.zst
2024-01-27T12:32:24.1638141Z INFO:root:Content length is 867901576 bytes
2024-01-27T12:32:34.5166393Z INFO:root:Downloading finished
2024-01-27T12:32:34.5207334Z INFO:root:Going to run command docker run --privileged --volume=/home/ubuntu/actions-runner/_work/_temp/test/result:/workspace --volume=/home/ubuntu/actions-runner/_work/_temp/test/result:/output --volume=/home/ubuntu/actions-runner/_work/ClickHouse/ClickHouse/tests:/usr/share/clickhouse-test --volume=/home/ubuntu/actions-runner/_work/_temp/test:/artifacts --cap-add syslog --cap-add sys_admin --cap-add sys_rawio -e CHECK_START_TIME='2024-01-27 12:31:54' -e CHECK_NAME='Performance Comparison' -e INSTANCE_TYPE='m6a.4xlarge' -e INSTANCE_ID='i-065d06469518b2dbc' -e PR_TO_TEST=59287 -e SHA_TO_TEST=44e42052b184834466970a7d95a3943eba572dec  -e S3_URL=https://s3.amazonaws.com/clickhouse-builds -e BUILD_NAME=package_release -e CHPC_ADD_REPORT_LINKS="<a href=https://github.com/ClickHouse/ClickHouse/actions/runs/7678532452>Job (actions)</a> <a href=https://github.com/ClickHouse/ClickHouse/pull/59287>Tested commit</a>" -e CHPC_TEST_RUN_BY_HASH_TOTAL=4 -e CHPC_TEST_RUN_BY_HASH_NUM=0 -e CLICKHOUSE_PERFORMANCE_COMPARISON_DATABASE_URL -e CLICKHOUSE_PERFORMANCE_COMPARISON_DATABASE_USER -e CLICKHOUSE_PERFORMANCE_COMPARISON_DATABASE_USER_PASSWORD -e CLICKHOUSE_PERFORMANCE_COMPARISON_CHECK_NAME -e CLICKHOUSE_PERFORMANCE_COMPARISON_CHECK_NAME_PREFIX clickhouse/performance-comparison:adc182c3de19
```

So odd.

Reply:
Thank you for the update!
I created WA for the issue here: #59325

Reply:
Likely this will fix the problem thank you.
What do you think about cleaning this up, i.e.
- move the code to download archives into one place, so that nobody will misunderstand this (right now there is code in `download.sh` and in python)
- remove other unused code (if any)
- I guess `performance.tgz` can be removed already (it is only possible in release branches and this change should not be backported) - https://github.com/ClickHouse/ClickHouse/blob/0f88da5fd4a0ef51c597944e49f9336119fef052/tests/performance/scripts/entrypoint.sh#L67-L69, since later there is a check only against `performance.tar.zst`
- and even maybe write some comments in the relevant parts of the code to show how it works

Reply:
Absolutely agree. Cleaning this up here is in my wish list for a few months already.
Perf test is somewhat cumbersome now, and, I believe, there are some things that could be simplified.
Your points look reasonable to me. I also want to dive into perf test a bit and check if maybe we do not need this _performance.tar.zst_ at all. With the current perf test design it's not very clear what's going on under the hood and it's frustrating when you have to work with it.